### DISCUSSION TITLE
Group regex fallbacks ( that follow a group sequence )

### DISCUSSION THREAD

**Discussion by JorgeR81** [chrisgoringe/cg-use-everywhere#168] [Ideas]
**_Summary_**

I'd like to be able to type something like: "_ControlNet>Sampler_", in the [ group_regex ] of a UE? node.

If the [ ControlNet ] group is available in the workflow, then the UE? link will go to this group. 
If the [ ControlNet ] group is muted, the UE? link will go to the "_Sampler_" groups, instead.

I will explain the problem in more detail, in case you can come up with a better solution.

_( there some are hidden images to help visualising it ... )_

----

**_Problem_**

In my workflow, I start with the main prompt / conditioning distributed with a [ Prompts Everywhere ] node.
Then I have a group named "_Combine Prompt_", where the main prompt is combined with another prompt. 
This group will send the conditioning to a "_ControlNet_" group, via UE?.
And, finally the _"ControlNet_" group, will send to "_Sampler_" groups, also via UE?.

The "_Combine Prompt_" and "_ControlNet_" groups are not essential, so they can be muted. 
But, if only the "_ControlNet_" is muted, the sequence breaks, and "_Combine Prompt_" is not sent to the samplers.
The workflow this works, but only with the basic prompt, form the [ Prompts Everywhere ] node.

So, I need to delete "ControlNet" and type "Sampler", in the UE? ( via Simple String node ), for [ Combine Prompt ] group, every time I mute the [ ControlNet ] group.
And If I enable the [ ControlNet ] group, I need to change it back again, every time. 

_Here are some images to help visualising it:_ 

<details>
<summary>Workflow with all groups enabled</summary>
<IMG src="https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/7b80dbb1-c579-4213-8dd8-06cb0ddae635"  alt="image.jpg"/>
</details>

<details>
<summary>Workflow with non essential groups muted</summary>
<IMG src="https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/8513f3f9-3f15-4194-bfbe-a15b1a27c1f2"  alt="image.jpg"/>
</details>

<details>
<summary>Combine Prompt muted</summary>
<IMG src="https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/28afd570-98c6-445f-bca6-71322679bbb3"  alt="image.jpg"/>
</details>

<details>
<summary>ControlNet muted  ( the sequence breaks )</summary>
<IMG src="https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/bc2c5a55-2f0d-4fb7-a08b-f534cf8e11c7"  alt="image.jpg"/>
</details>

<details>
<summary>ControlNet Muted and [ group_regex ] adjusted</summary>
<IMG src="https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/2dfbd967-109f-43f3-bec4-4673cc41121f"  alt="image.jpg"/>
</details>

---

**_Workaround_**

What I do now, is to bypass specific nodes, inside the non essential groups. 
The UE links are transmitted across these groups, but the group has no effect in the final image.
As you know, we can't bypass the whole group, because this also breaks UE links.

<details>
<summary>Workflow with non essential nodes bypassed</summary>
<IMG src="https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/127dfe1c-171d-437e-b7e9-27d57261e4ed"  alt="image.jpg"/>
</details>

---

_**Potential solution**_

I want to have a rule like this: "_CombineP>ControlNet>Sampler_" in the [ group_regex ], for the UE? nodes to follow. 

So, in the [ CombineP ] group, I would type: "_ControlNet>Sampler_", in the UE?  [ group_regex ] widget.

And if the [ ControlNet ] group is muted, the UE? node in the [ CombineP ] group, would send the link to "_Sampler_" instead.

This would also be useful for more complex workflows:

If I want or have have a rule like this: "_CombineP>ControlNet_1>ControlNet_2>Sampler_"

- In the [ CombineP ] group's UE? widget, I would type: "_ControlNet_1>ControlNet_2>Sampler_"

- In the [ ControlNet_1 ] group's UE? widget, I would  type: "_ControlNet_2>Sampler_"

- In the [ ControlNet_2 ] group's UE? widget, I will just type: "_Sampler_", ( like I would do now ).

---

**_Advantages of this solution_**

- We can just use the [ Fast Groups Muter ( rgthree ) ] node, without any additional adjustments in the workflow.  
This allows me to have a "control panel" for all the extra functionality, in the non essential groups.

- Muting groups is a better solution ( vs bypassing nodes ), because when we enable UE link visualization, we can have a more clear picture of what the workflow is doing.

- I can save these extra groups as a templates, and if I load them in a new workflow, they can connect automatically !

**Comment by chrisgoringe** [comment]
Note suggestion in https://github.com/chrisgoringe/cg-use-everywhere/issues/170

**Comment by JorgeR81** [comment]
I also thought a hierarchy of priorities could solve this.

But I was not entirely sure on how that would work exactly ...

Perhaps a new EU? node with an extra widget, for priorities ?

And in the widget I would set: 1, 2, 3 ..., according to the group's position in sequence ?

**Comment by chrisgoringe** [comment]
[note - crosspost with the one immediately above]

@JorgeR81 I love this idea - its generality appeals to me. But it may be harder to implement than you would think - I'll explain more below. Would @DrJKL 's idea (an explicit priority) solve your problem?

The complication:

Although it's easy to think of UE nodes as sending, the way it works is from the other end: unconnected inputs fetching. The code builds a list of available UE sources, and each unconnected node finds one that works for it - and if there is more than one, picks the one with the highest priority score. (https://github.com/chrisgoringe/cg-use-everywhere?tab=readme-ov-file#priorities).

So the logic "send to A, but if it is muted, send to B" is a slight mismatch. Not impossible by any means; the node would have to choose from its list of regexes based on some rule (maybe "if there are no unmuted groups matching regex A, use regex B ... C ... D"). 

Whereas "give this UE a priority of X" is hooking into the existing approach. Also more intuitive, perhaps?

So my first thought would be to add an optional priority.

**Comment by chrisgoringe** [comment]
Nodes currently have a priority - highest wins. The priority scores are:

|node|score|
|-|-|
|UE?|10|
|Prompts|5|
|UE|2|
|UE3|0|

These are modified with +1 for group restrictions, and +1 for colour restrictions.

Easy enough to add to the UE? node a "priority boost". I'd put it on the right-click menu, and it would just be a list of possible boosts (0, +1, +2, ...). I would reduce the group and color modifiers to 0.1, so that they are only ever used as a tie-break.

I can do this in ten minutes, I think - so why don't I do it, and you can see if it solves the problem!

**Comment by JorgeR81** [comment]
Sure, I will see if "priority boost" helps with that.

Even if it's not applicable to that workflow in particular, I will try another workflow configuration.
I think this feature will be useful in all sorts of situations.

I believe @DrJKL's problem is the same as mine.

It's about being able to mute one group in a sequence of groups connected by UE? links, without breaking the UE? sequence.

**Reply by chrisgoringe** [reply]
Yes, that problem is a little more complex. My way of thinking about it is this: 

4 groups,  in a workflow A->B->C->D.

The -> is a UE send.

If C is muted, we want A->B->D. 

But this can't be solved by giving C's send a higher priority, because that would cause B to draw from C when C wasn't muted.

I'll look at the group regex fallback a bit later!

**Comment by chrisgoringe** [comment]
Take a look at the version just pushed.

![priority](https://github.com/chrisgoringe/cg-use-everywhere/assets/4371971/1f7c660e-03ce-4614-9c9d-fb4321e828ca)

**Comment by JorgeR81** [comment]
"Priority Boost" is working !

This is very easy to use and it opens the door for a whole new type of workflow solutions !

I wonder if there could be a way to display the priority value on the node ...

**Comment by JorgeR81** [comment]
> It's about being able to mute one group in a sequence of groups connected by UE? links, without breaking the UE? sequence.

I think I've found a way to solve this problem with "priority boost" !

4 groups, in a workflow A->B->C->D.

I want to mute C, without breaking the sequence at B.
So I added a second set of UE? nodes, in the B group, that also send to D group. 
Normally these would clash with the UE? nodes in the C group.
But now I give a priority boost to the nodes in group C, and all seems to work !

_EDIT: @chrisgoringe suggested a way to further improve this method. See his answer below the images._

---
A = Basic Prompt
B = Combine prompt group 
C = ControlNet group
D = Samplers group ( not pictured )

Here C is muted, and B is still sent do D.

![boost](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/ce6f544a-186a-4b37-9d0c-cc0b50215ba6)

---

All groups can work together, by giving "priority boost" to the UE? nodes, in group C.  

The "Sampler" UE? nodes in group C, have priority over the ones in group B.

![boost 2](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/0451a744-c4fd-4656-b0b6-110f6e66d92e)

**Reply by chrisgoringe** [reply]
B can send to C or D, C can send to D with higher priority. So if C is muted, B->D, but if not, B->C->D. 

That's a really neat solution. And I think it generalises pretty well.

And I think it can be made even better using the magic of regex: Just one set of ? nodes in B, with a group regex `ControlNet|Sampler` to match both groups.

[regex101.com](https://regex101.com) is amazingly useful...

**Reply by JorgeR81** [reply]
`ControlNet|Sampler` works great !

This can cover larger sequences, easily.
